node-api: add support for UTF-8 and Latin-1 property keys

mertcanaltin opened the pull request:
This commit addresses issue #52979. It introduces new functions for optimized property key creation:

node_api_create_property_key_utf8
node_api_create_property_key_latin1
These functions help in creating property keys that allow more efficient property access when reused multiple times.

This commit also includes:

Updates to documentation to reflect the addition of the new functions.
New tests to ensure the correctness of the implementation.
Additionally, this work is related to the previous PR [node-api: refactor napi_set_property function for improved performance #50282](https://github.com/nodejs/node/pull/50282), which covered the utf16 case.

Trott commented on doc/api/n-api.md:
```diff
@@ -3134,6 +3134,30 @@ overhead in creating/storing strings with this method.
 The JavaScript `string` type is described in
 [Section 6.1.4][] of the ECMAScript Language Specification.
 
+> **Stability**: 1 - Experimental
```
This seems like a strange place to put a stability banner. I would expect it to appear with a section heading (and accompanying YAML) indicating what part of the API it's talking about. See, for example, from earlier in the doc:

```md

#### `node_api_create_property_key_utf16`
<!-- YAML
added:
  - v21.7.0
  - v20.12.0
-->
> Stability: 1 - Experimental
```

mertcanaltin replied:
thank you for suggestions, ı applied your comments

Trott commented on doc/api/n-api.md:
```diff
@@ -3134,6 +3134,30 @@ overhead in creating/storing strings with this method.
+flexible property names, especially when dealing with virtual module
+names or other scenarios where '\0' characters are used as part of the name.
+
+Note: This function is experimental and its stability may change in future releases.
```
Since there's a stability banner, this would seem to be superfluous. Maybe remove it?

legendecas reviewed (commented):
I would prefer adding `node_api_set_named_property(napi_env env, napi_value object, const char* utf8name, size_t name_length, napi_value value)` (which is the new naming pattern) and deprecating `napi_set_named_property` once the new one promoted to stable.

legendecas commented on src/js_native_api_v8.cc:
```diff
@@ -1718,6 +1718,32 @@ napi_status NAPI_CDECL node_api_create_property_key_utf16(napi_env env,
+                                                   const char* utf8name,
+                                                   size_t name_length,
+                                                   napi_value value) {
+  std::u16string utf16name;
```
Please use `CHECK_NEW_FROM_UTF8_LEN` instead of raw casting.

mertcanaltin replied:
thank you for suggestions, ı applied your comments

legendecas commented on test/js-native-api/test_string/test_string.c:
```diff
@@ -310,6 +310,58 @@ static napi_value TestPropertyKeyUtf16AutoLength(napi_env env,
                          auto_length);
 }
 
+static napi_value TestSetNamedPropertyLen(napi_env env, napi_callback_info info) {
```
These new test functions are not called. Read more about named property tests at https://github.com/nodejs/node/blob/main/test/js-native-api/test_object/test_object.c#L250 and https://github.com/nodejs/node/blob/main/test/js-native-api/test_object/test.js#L118.

legendecas reviewed (dismissed):
Thank you for working on this. Request for change on implementation as comments above.

gabrielschulhof commented on doc/api/n-api.md:
```diff
@@ -3134,6 +3134,35 @@ overhead in creating/storing strings with this method.
 The JavaScript `string` type is described in
 [Section 6.1.4][] of the ECMAScript Language Specification.
 
+#### `napi_set_named_property_len`
```
New APIs should have the prefix `node_api_*`, i.e. this should be renamed to `node_api_set_named_property_len`.

gabrielschulhof commented on doc/api/n-api.md:
```diff
@@ -3134,6 +3134,35 @@ overhead in creating/storing strings with this method.
+  Returns napi\_ok if the API succeeded.
+
+This API sets a named property on a JavaScript object, treating '\0'
+characters as values rather than terminators. This allows for more
+flexible property names, especially when dealing with virtual module
+names or other scenarios where '\0' characters are used as part of the name.
```
```suggestion
characters as values rather than terminators.
```

Do we need to point out specific use cases? Perhaps we should concentrate on documenting the new capability itself.

gabrielschulhof commented on src/js_native_api.h:
```diff
@@ -118,6 +118,16 @@ NAPI_EXTERN napi_status NAPI_CDECL node_api_create_property_key_utf16(
 #endif  // NAPI_EXPERIMENTAL
 
+#ifdef NAPI_EXPERIMENTAL
+#define NODE_API_EXPERIMENTAL_HAS_PROPERTY_KEYS
```
```suggestion
#define NODE_API_EXPERIMENTAL_HAS_SET_NAMED_PROPERTY_LEN
```

This value is less generic and thus more tied to the specific API we're adding.

gabrielschulhof commented:
@mertcanaltin we discussed this PR at today's Node-API team meeting. We have come to the conclusion that, although this new API adds convenience for developers, it actually encourages a pattern of runtime inefficiency by making it easy to create properties with keys for which the engine is unaware that they are not just regular strings but property keys, and thus it doesn't optimize them.

We believe it would be better to pivot this PR to creating additional variants of `node_api_create_property_key_utf16` for the other two string encodings because it would encourage add-on maintainers to create property keys that are treated efficiently by the engine, thus increasing the performance of code that has to create large numbers of objects.

In addition, it would be advisable for add-on maintainers to create a property key once, and then reuse the resulting `napi_value` in subsequent calls to `napi_set_property` and/or `napi_define_properties` in order to avoid repeatedly creating the same string.

This is certainly less convenient for those writing the code, however, it increases the performance of the resulting add-ons.

mertcanaltin commented:
> @mertcanaltin we discussed this PR at today's Node-API team meeting. We have come to the conclusion that, although this new API adds convenience for developers, it actually encourages a pattern of runtime inefficiency by making it easy to create properties with keys for which the engine is unaware that they are not just regular strings but property keys, and thus it doesn't optimize them.
> 
> 
> 
> We believe it would be better to pivot this PR to creating additional variants of `node_api_create_property_key_utf16` for the other two string encodings because it would encourage add-on maintainers to create property keys that are treated efficiently by the engine, thus increasing the performance of code that has to create large numbers of objects.
> 
> 
> 
> In addition, it would be advisable for add-on maintainers to create a property key once, and then reuse the resulting `napi_value` in subsequent calls to `napi_set_property` and/or `napi_define_properties` in order to avoid repeatedly creating the same string.
> 
> 
> 
> This is certainly less convenient for those writing the code, however, it increases the performance of the resulting add-ons.

First of all, thank you very much for your review and support, do you think I can continue to develop this pr I would be very happy if you have any suggestions ❤️🙏

mertcanaltin commented:
I created a node_api_create_property_key_utf8 based on what you said

vmoroz commented on src/js_native_api.h:
```diff
@@ -118,6 +118,22 @@ NAPI_EXTERN napi_status NAPI_CDECL node_api_create_property_key_utf16(
 #endif  // NAPI_EXPERIMENTAL
 
+#ifdef NAPI_EXPERIMENTAL
+#define NODE_API_EXPERIMENTAL_HAS_PROPERTY_KEYS
```
Defining the same macro twice will cause compiler warnings or errors. Please group the new `node_api_create_property_key_utf8` API with the previous `node_api_create_property_key_utf16`.

vmoroz commented on src/js_native_api_v8.cc:
```diff
@@ -1718,6 +1718,46 @@ napi_status NAPI_CDECL node_api_create_property_key_utf16(napi_env env,
   });
 }
 
+napi_status node_api_create_property_key_utf8(napi_env env,
```
Please implement the `node_api_create_property_key_utf8` the same way as the `node_api_create_property_key_utf16` implemented using the `v8impl::NewString` helper function.
The current implementation is missing a few checks and uses `v8::NewStringType::kNormal` instead of `v8::NewStringType::kInternalized` which differentiates V8 normal string creation from a V8 string used as a property key.

mertcanaltin replied:
I tried to apply it as you suggested, thank you very much

vmoroz commented on test/js-native-api/test_string/test_string.c:
```diff
@@ -310,6 +310,93 @@ static napi_value TestPropertyKeyUtf16AutoLength(napi_env env,
                          auto_length);
 }
 
+static napi_value TestPropertyKeyUtf8(napi_env env, napi_callback_info info) {
```
The new test methods must be exposed to JS in the end of this code file and then used from JS test code.

vmoroz commented:
> I wonder if we can get a performance as you say by making different string encodings?

@mertcanaltin , there are a few Node-API benchmarks here: https://github.com/nodejs/node/tree/main/benchmark/napi. A new one can be created to benchmark different ways to set and get property values.
Note that the true advantage from using `node_api_create_property_key_utf8` and `node_api_create_property_key_utf16` must be seen only when the same property key is reused for setting/getting a property multiple times.

KevinEady commented:
I am hesitant of adding these APIs, as the functionality to do what you want already exists using both `node_api_create_property_key_utf16()` and `napi_set_property()`. I do not follow the argument of "an extra napi call" as outlined in https://github.com/nodejs/node/issues/52979#issuecomment-2110430131, as the team discussed it in a Node-API meeting and decided the additional call would be minimal overhead.

Adding new APIs has more overreaching effects than just Node.js itself, as other JavaScript engines must implement the new APIs as well. 

Since this functionality already exists and can be performed in a trivial manner, I 👎 this addition.

mertcanaltin commented:
Hi @KevinEady,
Some might argue that the existing node_api_create_property_key_utf16 and napi_set_property functions are sufficient. However, these functions come with additional performance costs and increased code complexity. The new function presents a more efficient and effective solution. I would love to hear any suggestions you might have.

mhdawson commented:
We discussed in the Node-api team meeting again today, and consensus seems to be  to implement the suggestion in https://github.com/nodejs/node/pull/52984#issuecomment-2117935616, and only add the new  property key creation functions, and no others.

mertcanaltin commented:
greetings I'm very sorry I'm late getting back here, I sent a few edits today

vmoroz commented:
@mertcanaltin , thank you for making the progress with the PR.
Based on the discussions in Node-API meetings and the comments above:
- Could you please remove the new `napi_set_named_property_len` function. The team sees this function as redundant. Developers can implement a simple helper function in their code today by using `napi_create_string_utf8` and `napi_set_property` functions to allow using `\0` in the name. Since Node-API is implemented by other JS runtimes such as Deno and Bun, and on top of other JS engines such as Hermes, adding simple helper functions unnecessary increases the API surface.
- As an alternative to the original purpose of this PR, it would be great to retarget it to add the new `node_api_create_property_key_latin1` and `node_api_create_property_key_utf8` functions. These functions enable creation of optimized strings that can be reused for the efficient property access. Their implementation can follow the same set of changes as in your previous PR #50282.
- It would be great to add a simple benchmark test that shows that reusing a property name created by `node_api_create_property_key_utf8` is faster than the one created by `napi_create_string_utf8` function.

Please let me know if I can help with the implementation.

vmoroz commented on doc/api/n-api.md:
````diff
@@ -3134,6 +3134,57 @@ overhead in creating/storing strings with this method.
+
+```c
+napi_value key;
+napi_status status = node_api_create_property_key_utf8(env, "myKey", strlen("myKey"), &key);
````
For a zero-terminating string we recommend using `NAPI_AUTO_LENGTH` instead. We call the `strlen` internally in such case. The explicit length is only required if the string may have `\0` inside or we want to use a substring. These two parameters are in some sense similar to the `std::string_view` behavior.

mertcanaltin replied:
many thanks I tried to apply as you said

mertcanaltin commented:
I did these steps @vmoroz 

- Removed the `napi_set_named_property_len` function based on discussions in Node-API meetings.
- Added `node_api_create_property_key_latin1` and `node_api_create_property_key_utf8` functions for optimized property key creation.
- Updated the test file to include new tests for `node_api_create_property_key_latin1` and `node_api_create_property_key_utf8`.
- Implemented a benchmark test (`BenchmarkPropertyKeyUtf8`) to compare the performance of `node_api_create_property_key_utf8` with `napi_create_string_utf8`.
- The new benchmark test shows that reusing a property name created by `node_api_create_property_key_utf8` is faster than using `napi_create_string_utf8`.

mertcanaltin commented:
I wonder if you have a suggestion about test failures  @vmoroz

vmoroz commented:
> I wonder if you have a suggestion about test failures @vmoroz
The error says that the module is not found. It is possible that it is failed to build. I will check it out and see how it can be fixed.

vmoroz commented:
> I did these steps @vmoroz
> 
> * Removed the `napi_set_named_property_len` function based on discussions in Node-API meetings.
> * Added `node_api_create_property_key_latin1` and `node_api_create_property_key_utf8` functions for optimized property key creation.
> * Updated the test file to include new tests for `node_api_create_property_key_latin1` and `node_api_create_property_key_utf8`.
> * Implemented a benchmark test (`BenchmarkPropertyKeyUtf8`) to compare the performance of `node_api_create_property_key_utf8` with `napi_create_string_utf8`.
> * The new benchmark test shows that reusing a property name created by `node_api_create_property_key_utf8` is faster than using `napi_create_string_utf8`.

@mertcanaltin , are you sure that the latest commit in this PR has all the latest code?
- The code is missing the implementation for `node_api_create_property_key_latin1` and there are no tests or docs for it.
- Docs for `napi_set_named_property_len` are not removed.
- The benchmark test must be in the [benchmark/napi](https://github.com/nodejs/node/tree/main/benchmark/napi) folder. E.g. see the [string](https://github.com/nodejs/node/tree/main/benchmark/napi/string) test there.

vmoroz commented on src/js_native_api_v8.cc:
```diff
@@ -1718,6 +1718,54 @@ napi_status NAPI_CDECL node_api_create_property_key_utf16(napi_env env,
+  v8::HandleScope handle_scope(isolate);
+
+  v8::Local<v8::String> maybe_key;
+  if (!v8::String::NewFromUtf8(isolate,
```
Please use `v8impl::NewString` as it is done for the `node_api_create_property_key_utf16`. The same is for the `node_api_create_property_key_latin1`

mertcanaltin replied:
ı used, thank you

mertcanaltin commented:
greetings, I tried to verify the structure, but I got a Cannot find package error @vmoroz 

```cpp
➜  node git:(dev-52979) ✗ make test
ninja -C out/Release 
ninja: Entering directory `out/Release'
ninja: no work to do.
if [ ! -r node ] || [ ! -L node ]; then ln -fs out/Release/node node; fi
/Applications/Xcode.app/Contents/Developer/usr/bin/make -s tooltest
.....
----------------------------------------------------------------------
Ran 5 tests in 0.002s

OK
/Applications/Xcode.app/Contents/Developer/usr/bin/make -s test-doc
node:internal/modules/esm/resolve:210
  const resolvedOption = FSLegacyMainResolve(packageJsonUrlString, packageConfig.main, baseStringified);
                         ^

Error: Cannot find package '/Users/mert/Desktop/openSource/node/tools/doc/node_modules/rehype-raw/package.json' imported from /Users/mert/Desktop/openSource/node/tools/doc/generate.mjs
    at legacyMainResolve (node:internal/modules/esm/resolve:210:26)
    at packageResolve (node:internal/modules/esm/resolve:829:14)
    at moduleResolve (node:internal/modules/esm/resolve:915:18)
    at defaultResolve (node:internal/modules/esm/resolve:1123:11)
    at ModuleLoader.defaultResolve (node:internal/modules/esm/loader:557:12)
    at ModuleLoader.resolve (node:internal/modules/esm/loader:526:25)
    at ModuleLoader.getModuleJob (node:internal/modules/esm/loader:249:38)
    at ModuleJob._link (node:internal/modules/esm/module_job:126:49) {
  code: 'ERR_MODULE_NOT_FOUND'
}

Node.js v23.0.0-pre
make[2]: *** [out/doc/api/addons.html] Error 1
make[1]: *** [doc-only] Error 2
make: *** [test] Error 2
➜  node git:(dev-52979) ✗ 
```

mertcanaltin commented:
in addition, markdown formatter was getting  error, I corrected it by using ```make format-md```

vmoroz commented on src/js_native_api_v8.cc:
```diff
@@ -1718,6 +1718,55 @@ napi_status NAPI_CDECL node_api_create_property_key_utf16(napi_env env,
+    return napi_invalid_arg;
+  }
+
+  napi_status status = v8impl::NewString(
```
We should use `return` here as in the `node_api_create_property_key_utf16`. Checking the return status does not bring any value. In the end we just return what the `v8impl::NewString` returned anyway.

vmoroz commented on test/js-native-api/test_string/test.js:
```diff
@@ -130,6 +130,19 @@ assert.strictEqual(test_string.TestPropertyKeyUtf16AutoLength(str6), str6);
+assert.strictEqual(Object.keys(obj1).length, 1);
+assert.strictEqual(obj1['\0test'], 42);
+
+const obj2 = test_string.TestSetNamedPropertyLenAutoLength();
```
The tests for the TestSetNamedPropertyLen() and TestSetNamedPropertyLenAutoLength() must be removed as they are not exposed by the `test_string.c` anymore.
Instead, please try to follow the same test pattern as in the PR #50282

vmoroz commented on test/js-native-api/test_string/test_string.c:
```diff
@@ -310,6 +310,39 @@ static napi_value TestPropertyKeyUtf16AutoLength(napi_env env,
                          auto_length);
 }
 
+static napi_value TestPropertyKeyUtf8(napi_env env, napi_callback_info info) {
```
This function can be greatly simplified by simply calling the `TestOneByteImpl` function.
See the PR #50282 where you had `TestPropertyKeyUtf16` and `TestPropertyKeyUtf16AutoLength` that simply call `TestTwoByteImpl`. Here we can do the same: we should have four functions: `TestPropertyKeyUtf8`, `TestPropertyKeyUtf8AutoLength`, `TestPropertyKeyLatin1`, and `TestPropertyKeyLatin1AutoLength` that just call the `TestOneByteImpl` with the corresponding parameters.

vmoroz commented on test/js-native-api/test_string/test.js:
```diff
@@ -130,6 +130,19 @@ assert.strictEqual(test_string.TestPropertyKeyUtf16AutoLength(str6), str6);
 assert.strictEqual(test_string.Utf16Length(str6), 5);
 assert.strictEqual(test_string.Utf8Length(str6), 14);
 
+const str7 = 'propertyKey';
```
Please make the JS tests for the new property creation to follow the same pattern as in the PR #50282.

mertcanaltin commented:
thank you very much for your review I tried to apply what you said @vmoroz

mertcanaltin commented:
ı added this test key, but get undefined error 

```js

vmoroz commented:
There are a couple of issues:
- The new unit tests are failing because not all strings are valid Latin1. In this case we should remove the Latin1 tests for the `str6`. Note that the normal string creation also does test the Latin1 there.
- The Benchmark test fails because it has no native code. Please follow the same pattern as the existing https://github.com/nodejs/node/tree/main/benchmark/napi/string test where they create a native module, and then test the exported functions from it in JS. We cannot simply call C code from JS like it is written currently. We must always export functions from a native module.

mertcanaltin commented:
this is great information thank you very much I will apply it

mertcanaltin commented:
applied many thanks @vmoroz 🚀

jasnell commented on test/js-native-api/test_string/test.js:
```diff
@@ -130,6 +130,7 @@ assert.strictEqual(test_string.TestPropertyKeyUtf16AutoLength(str6), str6);
 assert.strictEqual(test_string.Utf16Length(str6), 5);
 assert.strictEqual(test_string.Utf8Length(str6), 14);
 
+
```
```suggestion
```

mertcanaltin commented:
many thanks for helping me solve the errors here @legendecas

legendecas commented on doc/api/n-api.md:
```diff
@@ -12,11 +12,11 @@ maintained as part of Node.js itself. This API will be Application Binary
 addons from changes in the underlying JavaScript engine and allow modules
 compiled for one major version to run on later major versions of Node.js without
-recompilation. The [ABI Stability][] guide provides a more in-depth explanation.
+recompilation. The \[ABI Stability]\[] guide provides a more in-depth explanation.
```
Please restore unrelated changes in the doc.

legendecas commented on src/node_util.cc:
```diff
@@ -1,393 +1,3556 @@
-#include "node_dotenv.h"
-#include "node_errors.h"
-#include "node_external_reference.h"
+#include <algorithm>
```
Please restore unrelated changes.

legendecas reviewed (commented):
Please checkout https://github.com/legendecas/node/tree/node-api/property-keys for example changes.

legendecas commented on benchmark/napi/napi_property_benchmark/binding.gyp:
```diff
@@ -0,0 +1,14 @@
+      ],
+      "cflags": ["-std=c++11"],
+      "include_dirs": [
+        "<!(node -e \"require('nan')\")"
```
This is node-api addon and should not depend on `nan`. I think the benchmark could be a follow up since the APIs added are experimental.

I created https://github.com/nodejs/node/pull/54012 for existing `node_api_create_property_key_utf16`.

legendecas commented on test/js-native-api/test_string/test_string.c:
```diff
@@ -440,4 +495,4 @@ napi_value Init(napi_env env, napi_value exports) {
   return exports;
 }
-EXTERN_C_END
+EXTERN_C_END
```
Missing EOF.

mertcanaltin commented:
lint unfortunately seems to want the previous arrangement  @legendecas

legendecas commented:
@nodejs/node-api please take a look again :)

vmoroz commented on doc/api/n-api.md:
````diff
@@ -3088,6 +3088,52 @@ The native string is copied.
+
+```c
+napi_status NAPI_CDECL node_api_create_property_key_latin1(napi_env env,
+                                                           const char16_t* str,
````
it must be char*

vmoroz commented on doc/api/n-api.md:
````diff
@@ -3118,18 +3164,36 @@ This API creates an optimized JavaScript `string` value from
+
+```c
+napi_status NAPI_CDECL node_api_create_property_key_utf8(napi_env env,
+                                                         const char16_t* str,
````
same here: it must be char*

vmoroz commented:
@mertcanaltin, thank you for driving this PR to a completion! There are only a couple tiny typos in the docs.
Just a few notes about the PR description:
- Consider to add a link to the previous related PR #50282. Just to indicate that the utf16 case was covered previously in that PR.
- Please remove "A benchmark test to compare the performance of node_api_create_property_key_utf8 with napi_create_string_utf8." I understand that it is going to be covered in different PR.
- Consider to rephrase "These functions help in creating property keys more efficiently by reusing the same property key multiple times." to something like "These functions help in creating property keys that allow more efficient property access when are reused multiple times."

mertcanaltin commented:
made the latest updates, thank you very much for your support I learnt great things while doing this 🚀

legendecas commented on src/js_native_api_v8.cc:
```diff
@@ -1706,6 +1706,28 @@ napi_status NAPI_CDECL node_api_create_external_string_utf16(
+                                              napi_value* result) {
+  return v8impl::NewString(env, str, length, result, [&](v8::Isolate* isolate) {
+    return v8::String::NewFromUtf8(
+        isolate, str, v8::NewStringType::kNormal, static_cast<int>(length));
```
```suggestion
        isolate, str, v8::NewStringType::kInternalized, static_cast<int>(length));
```

gabrielschulhof commented on doc/api/n-api.md:
```diff
@@ -3088,6 +3088,52 @@ The native string is copied.
+of the whole strings.
+
+If a new JavaScript string is intended to be used as a property key, then for
+some JavaScript engines it will be more efficient to use the function in this
```
```suggestion
some JavaScript engines it will be more efficient to use the functions in this
```

gabrielschulhof commented on doc/api/n-api.md:
```diff
@@ -3088,6 +3088,52 @@ The native string is copied.
+If a new JavaScript string is intended to be used as a property key, then for
+some JavaScript engines it will be more efficient to use the function in this
+section. Otherwise, use the `napi_create_string_utf8` or
+`node_api_create_external_string_utf8` series functions as there may be
```
Why? If the engine doesn't provide a benefit when calling this function for any given implementation of Node-API, then, for that implementation, these functions should fall back to just calling `napi_create_string_*`, which should not incur no more than the burden of an additional indirection.

vmoroz replied:
There is a slight overhead for creating the internalized string because besides allocating the space for the string we should search for a string with the same content in the hash table. Depending on the string size, memory allocator, and number of strings in the hash table, the internalized strings could be more expensive or less expensive than normal string.
I guess for the small strings which can benefit from the small memory allocators, the lookup in the hash table is usually more expensive.

gabrielschulhof replied:
In our meeting today we decided to update the documentation for all these APIs after we land this PR.

gabrielschulhof commented on doc/api/n-api.md:
```diff
@@ -3088,6 +3088,52 @@ The native string is copied.
+
+This API creates an optimized JavaScript `string` value from
+an ISO-8859-1-encoded C string to be used as a property key for objects.
+The native string is copied.
```
Maybe we should add that, in contrast with `napi_create_string_latin1`, subsequent calls to this function with the same `str` pointer may benefit from a speedup in the creation of the requested `napi_value`, depending on the engine.

mertcanaltin replied:
added, thank you very much

anonrig commented on src/js_native_api_v8.cc:
```diff
@@ -1706,6 +1706,30 @@ napi_status NAPI_CDECL node_api_create_external_string_utf16(
+  return v8impl::NewString(env, str, length, result, [&](v8::Isolate* isolate) {
+    return v8::String::NewFromUtf8(isolate,
+                                   str,
+                                   v8::NewStringType::kInternalized,
```
Why not kNormal?

mertcanaltin replied:
I applied this, thank you very much

vmoroz replied:
@anonrig , I wonder what type of issues you saw with the Internalized strings. It is the whole point of this PR to introduce a set of functions to create Internalized strings in addition to the normal strings which we already have.

anonrig commented on src/js_native_api_v8.cc:
```diff
@@ -1706,6 +1706,30 @@ napi_status NAPI_CDECL node_api_create_external_string_utf16(
+  return v8impl::NewString(env, str, length, result, [&](v8::Isolate* isolate) {
+    return v8::String::NewFromOneByte(isolate,
+                                      reinterpret_cast<const uint8_t*>(str),
+                                      v8::NewStringType::kInternalized,
```
Why not kNormal?

mertcanaltin replied:
hello thank you very much for your comment @anonrig 

I thought this would be better in terms of performance
Memory Savingand such as Fast Comparisons

jasnell replied:
If these values can be user controlled, then `kNormal` is likely better here.

mertcanaltin replied:
I applied this, thank you very much

vmoroz replied:
@jasnell , could you please provide more details on the issues with the Internalized strings.
The goal of these new APIs is to create strings to be used as property keys. The internalized strings must be more efficient because otherwise V8 must internalize Normal string every single time.

vmoroz commented on src/js_native_api_v8.cc:
```diff
@@ -1706,6 +1706,28 @@ napi_status NAPI_CDECL node_api_create_external_string_utf16(
+  return v8impl::NewString(env, str, length, result, [&](v8::Isolate* isolate) {
+    return v8::String::NewFromOneByte(isolate,
+                                      reinterpret_cast<const uint8_t*>(str),
+                                      v8::NewStringType::kNormal,
```
The whole point of these new functions is to use the internalized strings. We already have functions that create Normal strings. No need to duplicate them with different names.

vmoroz reviewed (changes requested):
We must ensure that the new APIs to create property keys use the Internalized strings. Otherwise, we already have functions to create normal strings. No need to duplicate them.

legendecas commented:
I agree that the key point of the new APIs is introducing internalized strings that can be used as optimized property names.

Jarred-Sumner commented:
Does V8 not use a reference-counted string type? 

For JavaScriptCore, the performant way to do this would be to _not_ make it a napi_value (or a JSValue) and instead be a pointer with a different type declaration that uses the existing reference-counted mechanism (`WTF::AtomString` and  `JSC::Identifier`). It's a bit complicated to expose though, because it would mean a ref function, an unref function, and separate functions for putting/getting properties which accept that type. On the other hand, these strings work across global objects within the same thread (`napi_env`)

mertcanaltin commented:
Thank you, Jarred, for your suggestion. I find the idea of using a reference-counted string mechanism in V8 interesting, and it could certainly offer performance benefits. However, I think there are some challenges to implementing such a change for Node-API at the moment:

Implementation Complexity: Reference-counted string management would need to be compatible with the existing napi_value infrastructure. The new system would also need to be supported by all JavaScript engines that Node.js works with, which could lead to compatibility issues.

Performance Gains: While reference-counted strings can offer performance advantages, we believe that our current napi_value and string management setup is sufficiently fast in certain scenarios. Specifically, when a string needs to be reused multiple times, we observe that the node_api_create_property_key_utf8 and node_api_create_property_key_utf16 functions provide adequate optimization.

Maintenance and Complexity: Adding a new string management mechanism could make the maintenance of the API more complex. Additional documentation and guidance might be necessary to ensure that the developer community adopts and uses this new system correctly.

We are open to further evaluating this proposal and collecting more data through performance testing to better understand the benefits of these approaches in real-world usage scenarios. Thank you again for your feedback, and we look forward to more discussions!

vmoroz commented:
> Does V8 not use a reference-counted string type?
> 
> For JavaScriptCore, the performant way to do this would be to _not_ make it a napi_value (or a JSValue) and instead be a pointer with a different type declaration that uses the existing reference-counted mechanism (`WTF::AtomString` and `JSC::Identifier`). It's a bit complicated to expose though, because it would mean a ref function, an unref function, and separate functions for putting/getting properties which accept that type. On the other hand, these strings work across global objects within the same thread (`napi_env`)

@Jarred-Sumner , no, V8 does not use ref counting anywhere. They rely on GC. They even offer GC for the native code that must work with V8.
So, you think current API cannot be efficiently implemented for JavaScriptCore Node-API, right?
Is there a way to maintain some kind of tracking of napi_value and related scopes that would internally change the ref count for the `WTF::AtomString` and `JSC::Identifier`? If you design such API, then how it would look to be efficient for JSC?

mertcanaltin commented:
I restored it to preserve the string type @vmoroz

jasnell commented:
@mertcanaltin ... can I ask you to manually squash and rebase this PR. It's not landing cleanly as is.

mertcanaltin commented:
> @mertcanaltin ... can I ask you to manually squash and rebase this PR. It's not landing cleanly as is.

did the rebase, many thanks for your review

jasnell commented:
PR is currently blocked from landing by unreliable CI.

jasnell commented:
This PR is currently blocked from landing due to unreliable CI.

jasnell commented:
Landed in 75e4d0df8f4e

legendecas commented:
@jasnell you got ahead of me in seconds... Thanks for landing it!
